snovault/schema_utils.py

```
@@ -115,13 +116,48 @@ def json_file_contains_element(json_filename: str, json_element: str) -> bool:
     return schema_ref
 
 
+def fetch_schema_by_package_name(package_name: str, schema_name: str) -> dict:
+    """ Uses the pkg_resources library (good through 3.11) to resolve schemas in other
+        packages """
+    try:
+        schema_data = pkg_resources.resource_string(package_name, schema_name)
+        schema = json.loads(schema_data)
+        return schema
+    except Exception as e:
+        raise RefResolutionError(f"Failed to fetch schema by package name: {str(e)}")
+
+
+def fetch_field_from_schema(schema: dict, ref_identifer: str):
+    """ Fetches a field from a schema given the format in our $merge definitions ie:
+        /properties/access_key_id -> schema.get('properties', {}).get('access_key_id', {})
+    """
+    split_path = ref_identifer.split('/')[1:]
+    resolved = None
+    for subpart in split_path:
+        resolved = schema.get(subpart, {})
+        schema = resolved
+    if not resolved:
+        raise RefResolutionError(f'Could not locate $merge ref {ref_identifer} in schema')
+    if not isinstance(resolved, dict):
+        raise ValueError(
+            f'Schema ref {ref_identifer} must resolve dict, not {type(resolved)}'
+        )
+    return resolved
+
+
 def resolve_merge_ref(ref, resolver):
-    with resolver.resolving(ref) as resolved:
-        if not isinstance(resolved, dict):
-            raise ValueError(
-                f'Schema ref {ref} must resolve dict, not {type(resolved)}'
-            )
-        return resolved
+    try:
+        with resolver.resolving(ref) as resolved:
+            if not isinstance(resolved, dict):
+                raise ValueError(
+                    f'Schema ref {ref} must resolve dict, not {type(resolved)}'
+                )
+            return resolved
+    except (RefResolutionError, ValueError):  # try again under a different method
```

Does the `ValueError` need to be caught here? If it resolved successfully, then I'm not sure what will change by resolving it in a different manner.

---

Both errors can be thrown legitimately in a fail to resolve scenario so I think they both need to be caught here

---

Just so I understand - you're just making sure that the thing it resolves to is a dictionary, right? And it must, otherwise it is not correctly resolved - it resolved to the wrong type, yes?

---

Yeah I guess I should remove the ValueError catch here as Doug has a point, its probably a legitimate reference that is "too deep", so should let it fail through